File: src/08.dbpointer/01.345.ts
export function reverseVowels(s: string): string {
    // 'a'、'e'、'i'、'o'、'u'

    // hello
    // eo
    // oe
    // holle

    // leetcode
    // eeoe
    // eoee
    // leotcede

    const arr = s.split("");
    const isVowels = (ch: string) => {
        return "aeiouAEIOU".includes(ch);
    };

    let i = 0;
    let j = arr.length - 1;

    while (i < j) {
        if (isVowels(arr[i]) && isVowels(arr[j])) {
            // 交换位置
            let temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            i++;
            j--;
            continue;
        }

        if (!isVowels(arr[i])) {
            i++;
        }

        if (!isVowels(arr[j])) {
            j--;
        }
    }

    return arr.join("");
}

File: src/08.dbpointer/01.345.go
package dbpointer

import "strings"

func ReverseVowels(s string) string {
	arr := []byte(s)
	vowels := "aeiouAEIOU"

	i := 0
	j := len(arr) - 1

	for i < j {
		isExistLeft := strings.ContainsRune(vowels, rune(arr[i]))
		isExistRight := strings.ContainsRune(vowels, rune(arr[j]))

		if isExistLeft && isExistRight {
			temp := arr[i]
			arr[i] = arr[j]
			arr[j] = temp
			i++
			j--
			continue
		}

		if !isExistLeft {
			i++
		}

		if !isExistRight {
			j--
		}
	}

	return string(arr)
}
